django-server/server/api/urls.py

```diff
@@ -4,6 +4,7 @@
 
 router = routers.DefaultRouter()
 router.register(r'cafeterias', views.CafeteriaViewSet)
+router.register(r'base_cafeterias', views.CafeteriaBaseViewSet)
```

Comment:
How does it differ from `/cafeterias`? Maybe we can just use filtering in the latter?

Reply:
Here somebody had the same problem, and apparently there is a package that does this automatically: https://stackoverflow.com/questions/23643204/django-rest-framework-dynamically-return-subset-of-fields

Reply:
having base cafeeria you can get all of them in short time while cafeterias consits of complicated json structure (cafeteria has menu a which have reviews and tags and all of the are downloaded using basic api call to cafeterias - > it's too much data to just show the list of cafeterias and  let user choose his one). Using the soecified solution from stack it may be harder to play with data in application when there is single 'universal' api naming because you have to keep track of the fields which are available in these objects on your own (some of them may be nullable and it atarts being annoying for a group of fields)

Reply:
Well, then it introduces mutual dependency between client and the API.

Imagine client introduces new feature. Then you have to modify the client (obviously) **and** the API to accommodate for this change. What if we have multiple versions of the application out in the wild? Then we may have to support multiple versions for this one client. The cost of maintaining the API increases, because there are multiple versions of nearly the same functionality. E.g. if there is a bug in one of the features used by Android too, then you have to fix android-specific method for all these versions, and all copies of a regular method too.

Now imagine there is a new iOS client with a slightly differing set of features. Well, we either use the general API or create a new `base_*` methods for iOS specifically. A new client for a desktop app? New copy of the methods, and they all come in different versions. That gives you multiple copies of all methods in all clients: `cafeterias`, `base_cafeterias_android`, `base_cafeterias_ios`, `base_cafeterias_winapi` etc. We now have four copies for each API version.

The problem with API may be solved not by copying methods, but by providing more specified methods. Menu is an object on its own, it should be served  :) separately. Instead of having `/cafeterias` provide all menu options, cameras and statistics we can have
```
/cafeterias
/cafeterias/{id}/menus
/cafeterias/{id}/devices
etc.
```
Particular client then makes only the requests that it needs.

This approach ensures there is no need for extra checks like `nullptr`s mentioned, because you fetch only the bits and pieces you need. Also, it helps to better utilize new caching mechanism.

Initially I misunderstood the problem, it's not about using a subset of fields (which may come in handy at some points), but rather breaking down existing methods.

Reply:
ok, it was harder to find out how to do it in a way visible for api generator and docs page but after tries of different libraries found out that there is one that fits the requirements. let's see the effect on https://eaterslab-dev.herokuapp.com/api/beta/cafeterias/1/fixed_menu_options/1/ and https://eaterslab-dev.herokuapp.com/doc